# Assignment 8

## Network Analysis

### Step 2: Take the Shakespeare play you've previously been analyzing and construct a network visualization

- If you've been analyzing Macbeth, choose another play -- I will ask you to re-do the assignment if you use the same example from the lecture notebooks
- You are welcome to use any or all of NetworkX, Bokeh, or Dash Cytoscape
- You are also free to determine the information from the play that you use for nodes and edges
- You must clarify (in markdown cells) what information is contained in the graph and how you are measuring it
- Also include a description (via markdown cells) of the network's density and the "most important" nodes
    - You can choose what "most important" means, but use a quantitative metric and include this metric's value(s) in your description
- Tailor the graph's aesthetics to enhance the visualization

In [1]:
import requests
import networkx as nx
import matplotlib.pyplot as plt

In [2]:
response = requests.get('https://www.gutenberg.org/ebooks/1531.txt.utf-8')
response

<Response [200]>

In [3]:
othello = response.text

In [4]:
characters = [
        'DUKE OF VENICE','BRABANTIO','GRATIANO','LODOVICO',
        'OTHELLO','CASSIO','IAGO','MONTANO',
        'RODERIGO','CLOWN','DESDEMONA','EMILIA','BIANCA'
]

In [5]:
#the number of acts in Othello
othello_acts = othello.split('ACT')[6:11]
len(othello_acts)

5

In [6]:
charnum = {}

connections = {}
for i in range(len(characters)-1):
    for j in range(i+1,len(characters)):
        connections[(characters[i],characters[j])] = 0

for k in characters:
    charnum[k] = 0

for i in othello_acts:
    for j in i.split('SCENE')[1:]:
        scenechars = []
        for k in characters:
            if j.find(k) != -1:
                scenechars.append(k)
                charnum[k] += 1
        for a in range(len(scenechars)-1):
            for b in range(a+1,len(scenechars)):
                connections[(scenechars[a],scenechars[b])] += 1

In [7]:
charnum

{'DUKE OF VENICE': 0,
 'BRABANTIO': 3,
 'GRATIANO': 2,
 'LODOVICO': 4,
 'OTHELLO': 12,
 'CASSIO': 9,
 'IAGO': 13,
 'MONTANO': 3,
 'RODERIGO': 7,
 'CLOWN': 2,
 'DESDEMONA': 9,
 'EMILIA': 8,
 'BIANCA': 3}

In [8]:
connections

{('DUKE OF VENICE', 'BRABANTIO'): 0,
 ('DUKE OF VENICE', 'GRATIANO'): 0,
 ('DUKE OF VENICE', 'LODOVICO'): 0,
 ('DUKE OF VENICE', 'OTHELLO'): 0,
 ('DUKE OF VENICE', 'CASSIO'): 0,
 ('DUKE OF VENICE', 'IAGO'): 0,
 ('DUKE OF VENICE', 'MONTANO'): 0,
 ('DUKE OF VENICE', 'RODERIGO'): 0,
 ('DUKE OF VENICE', 'CLOWN'): 0,
 ('DUKE OF VENICE', 'DESDEMONA'): 0,
 ('DUKE OF VENICE', 'EMILIA'): 0,
 ('DUKE OF VENICE', 'BIANCA'): 0,
 ('BRABANTIO', 'GRATIANO'): 0,
 ('BRABANTIO', 'LODOVICO'): 0,
 ('BRABANTIO', 'OTHELLO'): 2,
 ('BRABANTIO', 'CASSIO'): 1,
 ('BRABANTIO', 'IAGO'): 3,
 ('BRABANTIO', 'MONTANO'): 0,
 ('BRABANTIO', 'RODERIGO'): 3,
 ('BRABANTIO', 'CLOWN'): 0,
 ('BRABANTIO', 'DESDEMONA'): 1,
 ('BRABANTIO', 'EMILIA'): 0,
 ('BRABANTIO', 'BIANCA'): 0,
 ('GRATIANO', 'LODOVICO'): 2,
 ('GRATIANO', 'OTHELLO'): 2,
 ('GRATIANO', 'CASSIO'): 2,
 ('GRATIANO', 'IAGO'): 2,
 ('GRATIANO', 'MONTANO'): 1,
 ('GRATIANO', 'RODERIGO'): 1,
 ('GRATIANO', 'CLOWN'): 0,
 ('GRATIANO', 'DESDEMONA'): 1,
 ('GRATIANO', 'EMILIA'):

In [9]:
for k,v in charnum.items():
    print(k,v)

DUKE OF VENICE 0
BRABANTIO 3
GRATIANO 2
LODOVICO 4
OTHELLO 12
CASSIO 9
IAGO 13
MONTANO 3
RODERIGO 7
CLOWN 2
DESDEMONA 9
EMILIA 8
BIANCA 3


In [10]:
graphitems = []

for k,v in charnum.items():
    dashnode = {'data': {'id': k,
                         'label': k.title(),
                         'size': str(v)}}
    graphitems.append(dashnode)

for k,v in connections.items():
    if v != 0:
        dashedge = {'data': {'source': k[0],
                             'target': k[1],
                             'weight': v}}
        graphitems.append(dashedge)
    
graphitems

[{'data': {'id': 'DUKE OF VENICE', 'label': 'Duke Of Venice', 'size': '0'}},
 {'data': {'id': 'BRABANTIO', 'label': 'Brabantio', 'size': '3'}},
 {'data': {'id': 'GRATIANO', 'label': 'Gratiano', 'size': '2'}},
 {'data': {'id': 'LODOVICO', 'label': 'Lodovico', 'size': '4'}},
 {'data': {'id': 'OTHELLO', 'label': 'Othello', 'size': '12'}},
 {'data': {'id': 'CASSIO', 'label': 'Cassio', 'size': '9'}},
 {'data': {'id': 'IAGO', 'label': 'Iago', 'size': '13'}},
 {'data': {'id': 'MONTANO', 'label': 'Montano', 'size': '3'}},
 {'data': {'id': 'RODERIGO', 'label': 'Roderigo', 'size': '7'}},
 {'data': {'id': 'CLOWN', 'label': 'Clown', 'size': '2'}},
 {'data': {'id': 'DESDEMONA', 'label': 'Desdemona', 'size': '9'}},
 {'data': {'id': 'EMILIA', 'label': 'Emilia', 'size': '8'}},
 {'data': {'id': 'BIANCA', 'label': 'Bianca', 'size': '3'}},
 {'data': {'source': 'BRABANTIO', 'target': 'OTHELLO', 'weight': 2}},
 {'data': {'source': 'BRABANTIO', 'target': 'CASSIO', 'weight': 1}},
 {'data': {'source': 'BRABAN

In [11]:
import dash
import dash_cytoscape as cyto
from dash import html, dcc
from dash.dependencies import Input, Output
from jupyter_dash import JupyterDash
from jupyter_dash.comms import _send_jupyter_config_comm_request

In [12]:
_send_jupyter_config_comm_request()

In [14]:
JupyterDash.infer_jupyter_proxy_config()

In [15]:
app = JupyterDash(__name__)

app.layout = html.Div([
    cyto.Cytoscape(
        layout={'name': 'cose'},
        elements=graphitems,
        style={'width': '100%', 'height': '750px'},
        stylesheet=[
            {
                'selector': 'node',
                'style': {
                    'content':'data(label)',
                    'text-halign':'center',
                    'text-valign':'center',
                    'width': 'data(size)',
                    'height': 'data(size)',
                    'font-size':4,
                    'color': 'blue',
                    'text-outline-color': 'white',
                    'text-outline-width': 0.2,
                    'shape':'circle'
                }
            },
            {
                'selector':'edge',
                'style': {
                    'width':'data(weight)',
                    'line-color': 'black',
                  }
            },
        ]
    )
])

app.run_server(debug=True)

Dash app running on https://jupyter.idre.ucla.edu/user/crystal.huynh39@gmail.com/proxy/8050/


This visualization shows us that Iago has the most connections since his circle is the largest and his edges are relatively thicker than many other characters. This tells us that he is the most relevant character, which does make sense as in Assignment 6, his name was not only the most frequently mentioned out of all the characters, but it was also the word that appeared the most in the whole text. Othello seems to have the next most connections, then Cassio and Desdemona. The story of Othello revolves around these four characters the most, so this finding makes sense compared to the connections of other characters.

In [16]:
graphitems

[{'data': {'id': 'DUKE OF VENICE', 'label': 'Duke Of Venice', 'size': '0'}},
 {'data': {'id': 'BRABANTIO', 'label': 'Brabantio', 'size': '3'}},
 {'data': {'id': 'GRATIANO', 'label': 'Gratiano', 'size': '2'}},
 {'data': {'id': 'LODOVICO', 'label': 'Lodovico', 'size': '4'}},
 {'data': {'id': 'OTHELLO', 'label': 'Othello', 'size': '12'}},
 {'data': {'id': 'CASSIO', 'label': 'Cassio', 'size': '9'}},
 {'data': {'id': 'IAGO', 'label': 'Iago', 'size': '13'}},
 {'data': {'id': 'MONTANO', 'label': 'Montano', 'size': '3'}},
 {'data': {'id': 'RODERIGO', 'label': 'Roderigo', 'size': '7'}},
 {'data': {'id': 'CLOWN', 'label': 'Clown', 'size': '2'}},
 {'data': {'id': 'DESDEMONA', 'label': 'Desdemona', 'size': '9'}},
 {'data': {'id': 'EMILIA', 'label': 'Emilia', 'size': '8'}},
 {'data': {'id': 'BIANCA', 'label': 'Bianca', 'size': '3'}},
 {'data': {'source': 'BRABANTIO', 'target': 'OTHELLO', 'weight': 2}},
 {'data': {'source': 'BRABANTIO', 'target': 'CASSIO', 'weight': 1}},
 {'data': {'source': 'BRABAN

In [17]:
nxgraphitems = []

# for k,v in charnum.items():
#     dashnode = k
#     graphitems.append(dashnode)

for k,v in connections.items():
    if v != 0:
        dashedge = (k[0],k[1])
        nxgraphitems.append(dashedge)
    
nxgraphitems

[('BRABANTIO', 'OTHELLO'),
 ('BRABANTIO', 'CASSIO'),
 ('BRABANTIO', 'IAGO'),
 ('BRABANTIO', 'RODERIGO'),
 ('BRABANTIO', 'DESDEMONA'),
 ('GRATIANO', 'LODOVICO'),
 ('GRATIANO', 'OTHELLO'),
 ('GRATIANO', 'CASSIO'),
 ('GRATIANO', 'IAGO'),
 ('GRATIANO', 'MONTANO'),
 ('GRATIANO', 'RODERIGO'),
 ('GRATIANO', 'DESDEMONA'),
 ('GRATIANO', 'EMILIA'),
 ('GRATIANO', 'BIANCA'),
 ('LODOVICO', 'OTHELLO'),
 ('LODOVICO', 'CASSIO'),
 ('LODOVICO', 'IAGO'),
 ('LODOVICO', 'MONTANO'),
 ('LODOVICO', 'RODERIGO'),
 ('LODOVICO', 'DESDEMONA'),
 ('LODOVICO', 'EMILIA'),
 ('LODOVICO', 'BIANCA'),
 ('OTHELLO', 'CASSIO'),
 ('OTHELLO', 'IAGO'),
 ('OTHELLO', 'MONTANO'),
 ('OTHELLO', 'RODERIGO'),
 ('OTHELLO', 'CLOWN'),
 ('OTHELLO', 'DESDEMONA'),
 ('OTHELLO', 'EMILIA'),
 ('OTHELLO', 'BIANCA'),
 ('CASSIO', 'IAGO'),
 ('CASSIO', 'MONTANO'),
 ('CASSIO', 'RODERIGO'),
 ('CASSIO', 'CLOWN'),
 ('CASSIO', 'DESDEMONA'),
 ('CASSIO', 'EMILIA'),
 ('CASSIO', 'BIANCA'),
 ('IAGO', 'MONTANO'),
 ('IAGO', 'RODERIGO'),
 ('IAGO', 'CLOWN'),
 ('IA

In [18]:
G = nx.Graph()

In [19]:
G.add_edges_from(nxgraphitems)

In [20]:
G.nodes

NodeView(('BRABANTIO', 'OTHELLO', 'CASSIO', 'IAGO', 'RODERIGO', 'DESDEMONA', 'GRATIANO', 'LODOVICO', 'MONTANO', 'EMILIA', 'BIANCA', 'CLOWN'))

In [21]:
nx.degree_centrality(G)

{'BRABANTIO': 0.4545454545454546,
 'OTHELLO': 1.0,
 'CASSIO': 1.0,
 'IAGO': 1.0,
 'RODERIGO': 0.9090909090909092,
 'DESDEMONA': 1.0,
 'GRATIANO': 0.8181818181818182,
 'LODOVICO': 0.8181818181818182,
 'MONTANO': 0.7272727272727273,
 'EMILIA': 0.9090909090909092,
 'BIANCA': 0.8181818181818182,
 'CLOWN': 0.5454545454545454}

In [22]:
dc = nx.degree_centrality(G)
for i in sorted(dc, key=dc.get, reverse=True):
    print('{:15s}: {:.3f}'.format(i.title(), dc[i]))

Othello        : 1.000
Cassio         : 1.000
Iago           : 1.000
Desdemona      : 1.000
Roderigo       : 0.909
Emilia         : 0.909
Gratiano       : 0.818
Lodovico       : 0.818
Bianca         : 0.818
Montano        : 0.727
Clown          : 0.545
Brabantio      : 0.455


The degree centrality tells us that the most important nodes are Othello, Cassio, Iago, and Desdamonda. They all have the highest degree centrality at 1.0 which makes sense because again, the majority of the story revolves around the dynamic of these four characters the most. The next most important node is Rodrigo and Emilia o with a degree centrality of 0.909, which also makes sense because Rodrigo is in love with Desdemona, works with Iago, and tries to kill Cassio, thus placing him in the midst of their dynamic within the story. Emilia is Iago's wide and Desdemona's attendant, which also places her within their dynamic for much of the story.

### Step 3: When finished:

- Save a screenshot of your final network visualization (and/or save an html output file for Bokeh)
- Add both the saved screenshot/output-file and your notebook to your "dh140" repository on the JupyterHub
- Push the changes to your repository on GitHub
- Submit the link to your GitHub repository

Screenshot is saved under the name "Assignment08NetworkViz"